### PR TITLE
feat(cognito): advanced security option for userpool

### LINKED ISSUE #7405: Cognito: advanced security features for userpool
This is a followup issue from https://github.com/aws/aws-cdk/issues/7245#issuecomment-614658074

After creating a UserPool, the web console section that reads like the following:

> "Do you want to enable advanced security features for this user pool?"

Skärmavbild 2020-04-17 kl  21 01 40

Does not seem to be found on CDK.

### Use Case

I need this feature for security auditing purposes on a clinical genomics context.

### Proposed Solution

Find out the equivalent Cloudformation section for this, implement it in CDK.

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by winky**
### Commit Message

Support for  advanced security option in the user pool module. Advanced security option can be explicitly configured for enforced, audito or off

closes #7405

### End Commit Message

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 1]
```diff
@@ -322,6 +322,31 @@ export interface MfaSecondFactor {
...
+/**
+ * The different ways in which a user pool's Advanced Security Mode can be configured.
+ * @see https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-userpooladdons.html
+ */
+export enum AdvancedSecurityMode {
+  /** Enable advanced security */
+  ENFORCED = 'ENFORCED',
+  /** Gather information, and send user pool data to Amazon CloudWatch */
+  AUDIT = 'AUDIT',
+  /** Disenable advanced security */
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:346]:
```suggestion
  /** Disable advanced security */
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 2]
```diff
@@ -322,6 +322,31 @@ export interface MfaSecondFactor {
   readonly otp: boolean;
 }
 
+/**
+ * UserPooolAddOns settings for the user pool.
+ * @see https://docs.aws.amazon.com/ja_jp/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-userpooladdons.html
+ */
+export interface UserPoolAddOnsSettings {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:329]: Drop this structure and put the `advancedSecurityMode` directly into `UserPoolProps`.

File: packages/@aws-cdk/aws-cognito/README.md [hunk 3]
```diff
@@ -258,6 +258,21 @@ new UserPool(this, 'myuserpool', {
 
 Note that, `tempPasswordValidity` can be specified only in whole days. Specifying fractional days would throw an error.
 
+#### Advanced Security
+
+User ppols can be configured Advanced security. It can be turned on or used audit mode to gather metrics on detected risks without taking action. In audit mode, the advanced security features publishes metrics to Amazon CloudWatch. 
+See the [documentation on Advanced security](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pool-settings-advanced-security.html) to
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:264]: Looking at the documentation - https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pool-settings-advanced-security.html#cognito-user-pool-configure-advanced-security, once advanced security is set of 'ENFORCED', the user can configure a whole set of additional options.

I can't seem to find these options via CloudFormation. Are you aware of another way to set this up?
How useful is it to provide just the enum without the rest of the options?

Comment (winky) [reply packages/@aws-cdk/aws-cognito/README.md:264]: @nija-at This method (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpoolriskconfigurationattachment.html) is way to set this via CloudFormation.
Is it better to implement it in this PR.

Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/README.md:264]: I would say so, yes. 
As it is, it's not very useful to the customer to turn on advanced security but not have the option to configure it.

Comment (winky) [reply packages/@aws-cdk/aws-cognito/README.md:264]: I understand what you are saying.
I'm also going to implement an option to set that up.

Comment (winky) [reply packages/@aws-cdk/aws-cognito/README.md:265]: @nija-at I implement an option to setup user pool risk configuration. However, I'm not sure if it should be implemented in `user-pool.ts`, so please review it.

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 4]
```diff
@@ -681,6 +846,17 @@ export class UserPool extends Resource implements IUserPool {
...
+  /**
+   * Set risk Configuration for this user pool.
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pool-settings-advanced-security.html
+   */
+  public setUserPoolRiskConfiguration(id: string, options: UserPoolRiskConfigurationOptions): CfnUserPoolRiskConfigurationAttachment {
+    return new CfnUserPoolRiskConfigurationAttachment(this, id, {
+      userPoolId: this.userPoolId,
+      ...options,
+    });
+  }
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:858]: Separate this into its own construct called `UserPoolRiskConfigurationAttachment` and into a separate file called `user-pool-security.ts`. Move all associated interfaces into this file as well.

File: packages/@aws-cdk/aws-cognito/README.md [hunk 5]
```diff
@@ -258,6 +258,41 @@ new UserPool(this, 'myuserpool', {
...
+learn more.
+
+```ts
+const userPool = new UserPool(this, 'myuserpool', {
+  // ...
+  UserPoolAddOns: {
+    AdvancedSecurityMode: 'ENFORCED',
+  },
+});
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:274]:
```suggestion
const userPool = new UserPool(this, 'myuserpool', {
  // ...
  advancedSecurity: AdvancedSecurityMode.ENFORCED,
});
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 6]
```diff
@@ -474,6 +629,13 @@ export interface UserPoolProps {
    */
   readonly mfa?: Mfa;
 
+  /**
+   * Configure for advanced security mode .
+   *
+   * @default AdvancedSecurityMode.OFF.
+   */
+  readonly advancedSecurityMode?: AdvancedSecurityMode;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:637]:
```suggestion
  readonly advancedSecurity?: AdvancedSecurityMode;
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 7]
```diff
@@ -365,6 +378,148 @@ export interface PasswordPolicy {
...
+  /**
+   * The compromised credentials risk configuration including the event filter and action.
+   * @default - no configured
+   */
+  readonly compromisedCredentialsRiskConfiguration?: CompromisedCredentialsRiskConfiguration;
+  /**
+   * The configuration to override the risk decision.
+   * @default - no configured
+   */
+  readonly riskExceptionConfiguration?: CfnUserPoolRiskConfigurationAttachment.RiskExceptionConfigurationTypeProperty;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:520]: We don't expose the underlying Cfn resources here. Hide this under a custom modeled struct.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+  /**
+   * The compromised credentials risk configuration including the event filter and action.
+   * @default - no configured
+   */
+  readonly compromisedCredentialsRisk?: CompromisedCredentialsRiskConfiguration;
+  /**
+   * The configuration to override the risk decision.
+   * @default - no configured
+   */
+  readonly riskException?: RiskExceptionConfiguration;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:213]: Avoid this additional `RiskExceptionConfiguration` and directly put the two IP lists here?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+ */
+export interface RiskExceptionConfiguration {
+  /**
+   * Overrides the risk decision to always block the pre-authentication requests
+   */
+  readonly blockedIpRangeList?: string[];
+  /**
+   * Risk detection is not performed on the IP addresses in the range list
+   */
+  readonly skippedIpRangeList?: string[];
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:186]: Rename these to `alwaysBlockIpCidrs` and `alwaysAllowIpCidrs`. Call out in their documentation that these have to be in the CIDR format and provide a couple of examples.

**Comment by damienpuig-h4** [comment]
Hi! Is it possible to address all requested changes on this PR? would love to see this addition in aws-cognito

**Comment by huoxito** [comment]
@nija-at when you have a minute can you please confirm this feature is still not available in typescript aws-cdk? Is this something you could still approve? assuming we address the comments / feedback you made above

**Comment by nija-at** [comment]
> Is this something you could still approve? assuming we address the comments / feedback you made above

Yes, happy to approve this if the PR is cleaned up and addressed with the comments.

**Comment by ulaskeles** [comment]
@winky Have you had a chance to address the latest comments by @nija-at?

**Comment by ChrisSargent** [comment]
Hi - I was just looking for this functionality. Is there anything that's needed to be added to this PR such that it can be merged?

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+  /**
+   * The MFA email template used when MFA is challenged as part of a detected risk.
+   * @default - Not set
+   */
+  readonly mfaEmail?: NotifyEmail;
+  /**
+   * The email template used when a detected risk event is allowed.
+   * @default - Not set
+   */
+  readonly noActionEmail?: NotifyEmail;
```
[the remark this answers is not in the document]
Comment (pedrosola) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:98]: I found no reference on the [AWS Docs](https://docs.aws.amazon.com/cli/latest/reference/cognito-idp/set-risk-configuration.html)  for the behavior of notification when NotifyEmail is not set. It is likely that no email will be sent.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+  /**
+   * The subject of email.
+   */
+  readonly subject: string;
+  /**
+   * The text body of email.
+   * @default - Not set
+   */
+  readonly textBody?: string;
+}
```
[the remark this answers is not in the document]
Comment (pedrosola) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:73]: I found no reference on the AWS Docs for the behavior when both are set. I'll write a test to verify.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+  readonly noActionEmail?: NotifyEmail;
+  /**
+   * The destination to which the receiver of an email should reply to.
+   * @default - Not set
+   */
+  readonly replyTo?: string;
+  /**
+   * Email template used when a detected risk event is blocked.
+   */
+  readonly sourceArn: string;
```
[the remark this answers is not in the document]
Comment (pedrosola) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:107]: As per AWS Console: 
> Choose an email address that you have verified with Amazon SES.

This is a __SES Verified Identity__ Arn. Unsure how to proceed here as there is no construct for this resource.

File: packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+ */
+export interface NotifyConfiguration {
+  /**
+   * Email template used when a detected risk event is blocked.
+   * @default - Not set
+   */
+  readonly blockEmail?: NotifyEmail;
+  /**
+   * The email address that is sending the email.
+   * @default - Not set
```
[the remark this answers is not in the document]
Comment (pedrosola) [reply packages/@aws-cdk/aws-cognito/lib/user-pool-security.ts:86]: Documentation does not specify defaults. Will proceed to create a test for this functionality.

**Comment by TheRealAmazonKendra** [comment]
This PR has been deemed to be abandoned, and will be closed. Please create a new PR for these changes if you think this decision has been made in error.

**Pull request closed without merging**